policy/caa.go

```diff
@@ -0,0 +1,282 @@
+		// authority section of the answer.
+		if len(r.AnswerPacket.Ns) > 0 {
+			for _, record := range r.AnswerPacket.Ns {
+				if record.Header().Rrtype == dns.TypeSOA {
+					nsName = record.(*dns.SOA).Ns
+					rootDomain = record.(*dns.SOA).Hdr.Name
+					break
+				}
+			}
+			// silly ns in SOA...
+			if nsName == "." {
+				for _, record := range r.AnswerPacket.Ns {
+					if record.Header().Rrtype == dns.TypeSOA {
+						r, err := pubU.Resolve(record.(*dns.SOA).Hdr.Name, dns.TypeNS, dns.ClassINET)
+						if err != nil {
+							return "", "", fmt.Errorf("Resolving '%s' failed: %s", record.(*dns.SOA).Hdr.Name, err)
+						}
+						if r.HaveData {
+							for _, nsRecord := range r.Rr {
+								if nsRecord.Header().Rrtype == dns.TypeNS {
+									// just grab the first NS domain
+									nsName = nsRecord.(*dns.NS).Ns
+									rootDomain = nsRecord.(*dns.NS).Hdr.Name
+									break
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	if nsName == "" || rootDomain == "" {
+		return "", "", fmt.Errorf("Couldn't retrieve authoritative nameserver or zone root for '%s'", domain)
+	}
+	rootDomain = strings.TrimRight(rootDomain, ".")
+
+	nsIPs, err := pubU.LookupIP(nsName)
+	if err != nil {
+		return "", "", fmt.Errorf("Resolving '%s' failed: %s", nsName, err)
+	}
+	if len(nsIPs) > 0 {
+		// return first IP for NS
+		return nsIPs[0].String(), rootDomain, nil
+	}
+	return "", "", fmt.Errorf("Address lookup did not return any IPs for '%s'", nsName)
+}
+
+func getDNSKeys(pubU *unbound.Unbound, authU *unbound.Unbound, domain string, rootDomain string) (bool, error) {
+	splitDomain := strings.Split(rootDomain, ".")
+	// build Trust Anchor list
+	var taKeys []string
+	for i := range splitDomain {
+		taDomain := strings.Join(splitDomain[i:], ".")
+		r, err := pubU.Resolve(dns.Fqdn(taDomain), dns.TypeDNSKEY, dns.ClassINET)
+		if err != nil {
+			return false, fmt.Errorf("Resolving '%s' failed: %s", taDomain, err)
+		}
+		if r.HaveData {
+			for _, ta := range r.AnswerPacket.Answer {
+				if ta.Header().Rrtype == dns.TypeDNSKEY {
+					taKeys = append(taKeys, ta.String())
+				}
+			}
+		} else {
+			// cannot complete chain of trust, abort without error
+			return false, nil
+		}
+	}
+
+	// preload root key since every trust chain requires it
+	taKeys = append(taKeys, rootDNSKey)
+	for _, key := range taKeys {
+		if err := authU.AddTa(key); err != nil {
+			// only build *Chain of trust* if we have keys to make a full chain
+			// from the root domain to . (ie [google.com. -> com. -> .])
+			return false, fmt.Errorf("Couldn't add Trust Anchor: %s", err)
+		}
+	}
+
+	return true, nil
+}
+
+func getCaa(u *unbound.Unbound, domain string, alias bool) ([]*CAA, error) {
+	var CAAs []*CAA
+	if alias {
+		canonName, err := u.LookupCNAME(dns.Fqdn(domain))
+		if err != nil {
+			return CAAs, fmt.Errorf("CNAME lookup for '%s' failed: %s", domain, err)
+		}
+		if canonName == "" || canonName == domain {
+			return CAAs, nil
```

Comment:
Here I think it might be clearer to return &[]CAAs{}, assuming that works, to be clear that we intend to return an empty list.